flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/NonTimeUnboundedPrecedingFunction.java

```diff
@@ -0,0 +1,445 @@
+        rowData.setRowKind(RowKind.INSERT);
+        ListIterator<RowData> iterator = rowList.listIterator();
+
+        while (iterator.hasNext()) {
+            RowData curRow = iterator.next();
+            if (sortKeyComparator.compare(curRow, rowData) > 0) {
+                iterator.previous();
+                iterator.add(rowData);
+                isInserted = true;
+                break;
+            }
+            currFunction.accumulate(curRow);
+            prevFunction.accumulate(curRow);
+        }
+
+        // Add to the end of the list
+        if (!isInserted) {
+            iterator.add(rowData);
+        }
+
+        // Only accumulate rowData with currFunction
+        currFunction.accumulate(rowData);
+
+        // Update state with the newly inserted row
+        inputState.update(rowList);
+
+        // prepare output row
+        output.setRowKind(origRowKind);
+        output.replace(rowData, currFunction.getValue());
+        out.collect(output);
+
+        // Emit updated agg value for all records after newly inserted row
+        while (iterator.hasNext()) {
+            RowData curRow = iterator.next();
+            currFunction.accumulate(curRow);
+            prevFunction.accumulate(curRow);
+            // Generate UPDATE_BEFORE
+            output.setRowKind(RowKind.UPDATE_BEFORE);
+            output.replace(curRow, prevFunction.getValue());
+            out.collect(output);
+            // Generate UPDATE_AFTER
+            output.setRowKind(RowKind.UPDATE_AFTER);
+            output.replace(curRow, currFunction.getValue());
+            out.collect(output);
+        }
+    }
+
+    private void insertIntoSortedListOptimized(RowData input, Collector<RowData> out)
+            throws Exception {
+        List<RowData> sortedKeyList = sortedKeyState.value();
+        if (sortedKeyList == null) {
+            sortedKeyList = new ArrayList<>();
+        }
+        Long id = idState.value();
+        if (id == null) {
+            id = 0L;
+        }
+        boolean isInserted = false;
+        RowKind origRowKind = input.getRowKind();
+        input.setRowKind(RowKind.INSERT);
+        ListIterator<RowData> iterator = sortedKeyList.listIterator();
+
+        while (iterator.hasNext()) {
+            RowData curKey = iterator.next(); // (ID, sortKey)
+            RowData inputKey = GenericRowData.of(-1L, sortKeyFieldGetter.getFieldOrNull(input));
+
+            if (newSortKeyComparator.compare(curKey, inputKey) > 0) {
+                iterator.previous();
+                iterator.add(GenericRowData.of(id, sortKeyFieldGetter.getFieldOrNull(input)));
+                valueMapState.put(id, input);
+                isInserted = true;
+                id++;
+                break;
+            }
+            // Can also add the accKey to the sortedKeyList to avoid reading from the valueMapState
+            RowData curRow = valueMapState.get(curKey.getLong(keyIdx));
```

Comment:
I just realized that the Map approach won't helps if we can have to access all elements for any update (insert or delete). We can either persist the accumulators for all input rows (so we only need to recompute all accumulators after the inserted/deleted row) or if we split the input row into the fields that are needed for the aggregate computation and those that aren't